src/shared/clients/defaultS3Client.ts

```diff
@@ -0,0 +1,411 @@
+     * Looks up the region for the given bucket
+     *
+     * Note that although there is an S3#GetBucketLocation API,
+     * this is the suggested method of obtaining the region.
+     *
+     * @throws S3Error if there is an error calling S3.
+     */
+    private async lookupRegion(bucketName: string, s3: S3): Promise<string> {
+        try {
+            const response = await s3.headBucket({ Bucket: bucketName }).promise()
+            return response.$response.httpResponse.headers[BUCKET_REGION_HEADER]
+        } catch (e) {
+            throw new S3Error(`Failed to find region for bucket ${bucketName}: ${e}`)
+        }
+    }
+}
+
+export interface FileStreams {
+    createReadStream(uri: vscode.Uri): stream.Readable
+    createWriteStream(uri: vscode.Uri): stream.Writable
+}
+
+export class DefaultBucket implements Bucket {
+    public readonly name: string
+    public readonly region: string
+    public readonly arn: string
+
+    public constructor({ partitionId, region, name }: { partitionId: string; region: string; name: string }) {
+        this.name = name
+        this.region = region
+        this.arn = `arn:${partitionId}:s3:::${name}`
+    }
+
+    public toString(): string {
+        return `Bucket (name=${this.name}, region=${this.region}, arn=${this.arn})`
+    }
+}
+
+export class DefaultFolder implements Folder {
+    public readonly name: string
+    public readonly path: string
+    public readonly arn: string
+
+    public constructor({ partitionId, bucketName, path }: { partitionId: string; bucketName: string; path: string }) {
+        this.path = path
+        this.arn = `arn:${partitionId}:s3:::${bucketName}/${path}`
+        this.name = _(this.path)
+            .split(DEFAULT_DELIMITER)
+            .dropRight()!
+            .last()!
+    }
+
+    public toString(): string {
+        return `Folder (name=${this.name}, path=${this.path}, arn=${this.arn})`
+    }
+}
+
+export class DefaultFile implements File {
+    public readonly name: string
+    public readonly key: string
+    public readonly arn: string
+    public readonly lastModified: Date | undefined
+    public readonly sizeBytes: number | undefined
+
+    public constructor({
+        partitionId,
+        bucketName,
+        key,
+        lastModified,
+        sizeBytes,
+    }: {
+        partitionId: string
+        bucketName: string
+        key: string
+        lastModified: Date | undefined
+        sizeBytes: number | undefined
+    }) {
+        this.name = _(key)
+            .split(DEFAULT_DELIMITER)
+            .last()!
+        this.key = key
+        this.arn = `arn:${partitionId}:s3:::${bucketName}/${key}`
+        this.lastModified = lastModified
+        this.sizeBytes = sizeBytes
+    }
+
+    public toString(): string {
+        return `File (name=${this.name}, key=${this.key}, arn=${this.arn}, lastModified=${this.lastModified}, sizeBytes=${this.sizeBytes})`
+    }
+}
+
+class DefaultFileStreams implements FileStreams {
```

Comment:
I thought it was best to encapsulate the details of how the file is read inside of the client rather than requiring the caller to provide the stream. That way, this API becomes "uploadFile" instead of "uploadStream" which seems more intuitive and reusable. I didn't find the Node Streams that *fun* to work with, to put it lightly 😉.

I do agree that ideally the Uri param wouldn't be coupled to VSCode's stuff. However, the reason I did that was to prevent bugs in passing the wrong "format" of Uri. e.g. Windows vs Linux.

The VSCode Uri is a nice abstraction that hides all of those details. Only this class needs to know how translate that Uri into a string to feed it into Node file streams.

I'd apply a similar argument for using Date types instead of numbers that represent unix time  or something. An object that can have its own specific behavior, prevent bugs, and make interfaces easier to understand.